File: .project-starter/templates/health-check/python-fastapi/health.py
"""
Health Check Endpoints

Provides liveness and readiness endpoints for monitoring and orchestration.

Usage:
    from health import router
    app.include_router(router)
"""

from fastapi import APIRouter, status, Response
from pydantic import BaseModel
from typing import Optional
import logging
from sqlalchemy import text
from sqlalchemy.ext.asyncio import AsyncSession
# from redis import asyncio as aioredis  # Uncomment if using Redis

logger = logging.getLogger(__name__)

router = APIRouter(prefix="/health", tags=["health"])


class HealthChecks(BaseModel):
    """Detailed health checks for dependencies"""
    database: str
    cache: Optional[str] = None


class HealthResponse(BaseModel):
    """Health check response structure"""
    status: str
    service: Optional[str] = None
    version: Optional[str] = None
    checks: Optional[HealthChecks] = None


@router.get("", response_model=HealthResponse, status_code=status.HTTP_200_OK)
async def health_check():
    """
    Liveness probe endpoint

    Returns 200 if the service is running.
    Use this for Kubernetes liveness probes.

    **Response:**
    ```json
    {
        "status": "healthy",
        "service": "my-service",
        "version": "0.1.0"
    }
    ```
    """
    return HealthResponse(
        status="healthy",
        service="my-service",  # Replace with actual service name
        version="0.1.0"  # Replace with actual version
    )


@router.get("/ready", response_model=HealthResponse)
async def ready_check(
    db: AsyncSession,
    # redis: aioredis.Redis,  # Uncomment if using Redis
    response: Response
):
    """
    Readiness probe endpoint

    Returns 200 if the service is ready to accept traffic.
    Checks database and cache connections.
    Use this for Kubernetes readiness probes and load balancer health checks.

    **Response (healthy):**
    ```json
    {
        "status": "ready",
        "checks": {
            "database": "ok",
            "cache": "ok"
        }
    }
    ```

    **Response (unhealthy):**
    ```json
    {
        "status": "unhealthy",
        "checks": {
            "database": "error"
        }
    }
    ```
    Returns 503 status code if any check fails.
    """
    checks = {}
    is_healthy = True

    # Check database connection
    try:
        await db.execute(text("SELECT 1"))
        checks["database"] = "ok"
    except Exception as e:
        logger.error(f"Database health check failed: {e}")
        checks["database"] = "error"
        is_healthy = False

    # Optional: Check Redis connection
    # Uncomment if using Redis
    """
    try:
        await redis.ping()
        checks["cache"] = "ok"
    except Exception as e:
        logger.error(f"Redis health check failed: {e}")
        checks["cache"] = "error"
        is_healthy = False
    """

    if not is_healthy:
        response.status_code = status.HTTP_503_SERVICE_UNAVAILABLE
        return HealthResponse(
            status="unhealthy",
            service="my-service",
            version="0.1.0",
            checks=HealthChecks(**checks)
        )

    return HealthResponse(
        status="ready",
        service="my-service",
        version="0.1.0",
        checks=HealthChecks(**checks)
    )


# Example dependency injection setup
"""
from fastapi import Depends
from sqlalchemy.ext.asyncio import AsyncSession
from database import get_db

@router.get("/ready")
async def ready_check(
    db: AsyncSession = Depends(get_db),
    response: Response = None
):
    # Implementation above
    pass
"""

File: .project-starter/templates/health-check/nodejs-express/health.ts
/**
 * Health Check Endpoints
 *
 * Provides liveness and readiness endpoints for monitoring and orchestration.
 *
 * Usage:
 * ```typescript
 * import { healthRouter } from './health';
 * app.use('/health', healthRouter);
 * ```
 */

import { Router, Request, Response } from 'express';
import { Pool } from 'pg';
// import { createClient, RedisClientType } from 'redis';  // Uncomment if using Redis

const router = Router();

/**
 * Health check response structure
 */
interface HealthResponse {
  status: string;
  service?: string;
  version?: string;
  checks?: {
    database?: string;
    cache?: string;
  };
}

/**
 * Liveness probe endpoint
 *
 * Returns 200 if the service is running.
 * Use this for Kubernetes liveness probes.
 *
 * GET /health
 *
 * Response:
 * ```json
 * {
 *   "status": "healthy",
 *   "service": "my-service",
 *   "version": "0.1.0"
 * }
 * ```
 */
router.get('/', (req: Request, res: Response) => {
  const response: HealthResponse = {
    status: 'healthy',
    service: process.env.npm_package_name || 'my-service',
    version: process.env.npm_package_version || '0.1.0',
  };

  res.status(200).json(response);
});

/**
 * Readiness probe endpoint
 *
 * Returns 200 if the service is ready to accept traffic.
 * Checks database and cache connections.
 * Use this for Kubernetes readiness probes and load balancer health checks.
 *
 * GET /health/ready
 *
 * Response (healthy):
 * ```json
 * {
 *   "status": "ready",
 *   "checks": {
 *     "database": "ok",
 *     "cache": "ok"
 *   }
 * }
 * ```
 *
 * Response (unhealthy):
 * ```json
 * {
 *   "status": "unhealthy",
 *   "checks": {
 *     "database": "error"
 *   }
 * }
 * ```
 * Returns 503 status code if any check fails.
 */
router.get('/ready', async (req: Request, res: Response) => {
  const checks: { database?: string; cache?: string } = {};
  let isHealthy = true;

  // Get database pool from app locals (set during app initialization)
  const dbPool: Pool = req.app.locals.dbPool;

  // Check database connection
  try {
    await dbPool.query('SELECT 1');
    checks.database = 'ok';
  } catch (error) {
    console.error('Database health check failed:', error);
    checks.database = 'error';
    isHealthy = false;
  }

  // Optional: Check Redis connection
  // Uncomment if using Redis
  /*
  const redisClient: RedisClientType = req.app.locals.redisClient;
  try {
    await redisClient.ping();
    checks.cache = 'ok';
  } catch (error) {
    console.error('Redis health check failed:', error);
    checks.cache = 'error';
    isHealthy = false;
  }
  */

  const response: HealthResponse = {
    status: isHealthy ? 'ready' : 'unhealthy',
    service: process.env.npm_package_name || 'my-service',
    version: process.env.npm_package_version || '0.1.0',
    checks,
  };

  const statusCode = isHealthy ? 200 : 503;
  res.status(statusCode).json(response);
});

export { router as healthRouter };

/**
 * Example setup in main app file:
 *
 * ```typescript
 * import express from 'express';
 * import { Pool } from 'pg';
 * import { healthRouter } from './health';
 *
 * const app = express();
 *
 * // Initialize database pool
 * const dbPool = new Pool({
 *   connectionString: process.env.DATABASE_URL,
 * });
 *
 * // Store in app locals for health check access
 * app.locals.dbPool = dbPool;
 *
 * // Mount health check router
 * app.use('/health', healthRouter);
 *
 * // Optional: Redis setup
 * // import { createClient } from 'redis';
 * // const redisClient = createClient({ url: process.env.REDIS_URL });
 * // await redisClient.connect();
 * // app.locals.redisClient = redisClient;
 * ```
 */
